File: backend/app/unit_conversion.py
def _unit(unit: str | None) -> str:
    return (unit or "").strip().lower()


def glucose_to_mgdl(value, unit="mg/dL"):
    if value is None:
        return None
    if "mmol" in _unit(unit):
        return value * 18.0
    return value


def triglycerides_to_mgdl(value, unit="mg/dL"):
    if value is None:
        return None
    if "mmol" in _unit(unit):
        return value * 88.57
    return value


def cholesterol_to_mgdl(value, unit="mg/dL"):
    if value is None:
        return None
    if "mmol" in _unit(unit):
        return value * 38.67
    return value


def creatinine_to_mgdl(value, unit="mg/dL"):
    if value is None:
        return None
    normalized = _unit(unit)
    if "umol" in normalized or "µmol" in normalized:
        return value / 88.4
    return value


def bilirubin_to_mgdl(value, unit="mg/dL"):
    if value is None:
        return None
    normalized = _unit(unit)
    if "umol" in normalized or "µmol" in normalized:
        return value / 17.1
    return value


def albumin_to_gdl(value, unit="g/dL"):
    if value is None:
        return None
    if _unit(unit) == "g/l":
        return value / 10.0
    return value


def total_protein_to_gdl(value, unit="g/dL"):
    if value is None:
        return None
    if _unit(unit) == "g/l":
        return value / 10.0
    return value


def height_to_cm(value=None, unit="cm", feet=None, inches=None):
    if unit == "ft-in":
        if feet is None and inches is None:
            return None
        return (((feet or 0) * 12) + (inches or 0)) * 2.54
    if value is None:
        return None
    if "inch" in _unit(unit):
        return value * 2.54
    if "ft" in _unit(unit):
        return value * 30.48
    return value


def weight_to_kg(value, unit="kg"):
    if value is None:
        return None
    if _unit(unit) == "lb":
        return value * 0.453592
    return value


def waist_to_cm(value, unit="cm"):
    if value is None:
        return None
    if "inch" in _unit(unit):
        return value * 2.54
    return value


def temperature_to_c(value, unit="°C"):
    if value is None:
        return None
    normalized = _unit(unit)
    if normalized in ["°f", "f"]:
        return (value - 32.0) * 5.0 / 9.0
    return value


def platelets_to_10e9_l(value, unit="10⁹/L"):
    if value is None:
        return None
    normalized = _unit(unit)
    if "lakh" in normalized:
        return value * 100.0
    if "cells" in normalized:
        return value / 1000.0
    return value


def wbc_to_10e9_l(value, unit="10⁹/L"):
    if value is None:
        return None
    if "cells" in _unit(unit):
        return value / 1000.0
    return value


def blood_urea_to_mgdl(value, unit="mg/dL"):
    if value is None:
        return None
    if "mmol" in _unit(unit):
        return value * 6.006
    return value


def uric_acid_to_mgdl(value, unit="mg/dL"):
    if value is None:
        return None
    normalized = _unit(unit)
    if "umol" in normalized or "µmol" in normalized:
        return value / 59.48
    return value


def is_percent_unit(unit) -> bool:
    return "%" in _unit(unit)


def is_absolute_count_unit(unit) -> bool:
    normalized = _unit(unit)
    return "10" in normalized or "cells" in normalized


def absolute_count_to_10e9_l(value, unit):
    if value is None:
        return None
    if "cells" in _unit(unit):
        return value / 1000.0
    return value

File: frontend/lib/utils/unit_conversion.dart
double? glucoseToMgdl(double? value, String unit) {
  if (value == null) return null;
  if (_unit(unit).contains('mmol')) return value * 18.0;
  return value;
}

double? triglyceridesToMgdl(double? value, String unit) {
  if (value == null) return null;
  if (_unit(unit).contains('mmol')) return value * 88.57;
  return value;
}

double? cholesterolToMgdl(double? value, String unit) {
  if (value == null) return null;
  if (_unit(unit).contains('mmol')) return value * 38.67;
  return value;
}

double? creatinineToMgdl(double? value, String unit) {
  if (value == null) return null;
  if (_unit(unit).contains('umol') || _unit(unit).contains('µmol')) {
    return value / 88.4;
  }
  return value;
}

double? bilirubinToMgdl(double? value, String unit) {
  if (value == null) return null;
  if (_unit(unit).contains('umol') || _unit(unit).contains('µmol')) {
    return value / 17.1;
  }
  return value;
}

double? albuminToGdl(double? value, String unit) {
  if (value == null) return null;
  if (_unit(unit) == 'g/l') return value / 10.0;
  return value;
}

double? totalProteinToGdl(double? value, String unit) {
  if (value == null) return null;
  if (_unit(unit) == 'g/l') return value / 10.0;
  return value;
}

double? heightToCm(double? value, String unit) {
  if (value == null) return null;
  if (_unit(unit).contains('inch')) return value * 2.54;
  if (_unit(unit).contains('ft')) return value * 30.48;
  return value;
}

double? heightFeetInchesToCm(double? feet, double? inches) {
  if (feet == null && inches == null) return null;
  final totalInches = ((feet ?? 0) * 12) + (inches ?? 0);
  return totalInches * 2.54;
}

double? weightToKg(double? value, String unit) {
  if (value == null) return null;
  if (_unit(unit) == 'lb') return value * 0.453592;
  return value;
}

double? waistToCm(double? value, String unit) {
  if (value == null) return null;
  if (_unit(unit).contains('inch')) return value * 2.54;
  return value;
}

double? temperatureToC(double? value, String unit) {
  if (value == null) return null;
  if (_unit(unit) == '°f' || _unit(unit) == 'f') {
    return (value - 32.0) * 5.0 / 9.0;
  }
  return value;
}

double? plateletsTo10e9L(double? value, String unit) {
  if (value == null) return null;
  final normalized = _unit(unit);
  if (normalized.contains('lakh')) return value * 100.0;
  if (normalized.contains('cells')) return value / 1000.0;
  return value;
}

double? wbcTo10e9L(double? value, String unit) {
  if (value == null) return null;
  if (_unit(unit).contains('cells')) return value / 1000.0;
  return value;
}

double? bloodUreaToMgdl(double? value, String unit) {
  if (value == null) return null;
  if (_unit(unit).contains('mmol')) return value * 6.006;
  return value;
}

double? uricAcidToMgdl(double? value, String unit) {
  if (value == null) return null;
  if (_unit(unit).contains('umol') || _unit(unit).contains('µmol')) {
    return value / 59.48;
  }
  return value;
}

double? electrolyteToMmolL(double? value, String unit) {
  if (value == null) return null;
  return value;
}

double? absoluteCountTo10e9L(double? value, String unit) {
  if (value == null) return null;
  if (_unit(unit).contains('cells')) return value / 1000.0;
  return value;
}

bool isAbsoluteCountUnit(String unit) {
  final normalized = _unit(unit);
  return normalized.contains('10') || normalized.contains('cells');
}

bool isPercentUnit(String unit) => _unit(unit).contains('%');

String _unit(String unit) => unit.trim().toLowerCase();
